File: Day 015/Problem 148. Sort List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* merge(ListNode* left, ListNode* right) {
        ListNode* dummy = new ListNode(0);
        ListNode* ptr = dummy;

        while (left && right) {
            if (left->val < right->val) {
                ptr->next = left;
                left = left->next;
            } else {
                ptr->next = right;
                right = right->next;
            }
            ptr = ptr->next;
        }

        if (left) ptr->next = left;
        
        if (right) ptr->next = right;
        
        return dummy->next;
    }

    void mergeSort(ListNode *&head) {
        if (!head || !head->next) return;

        ListNode* slow = head, *fast = head, *supaSlow = nullptr;

        while (fast && fast->next) {
            supaSlow = slow;
            slow = slow->next;
            fast = fast->next->next;
        }

        supaSlow->next = nullptr;
        mergeSort(head);
        mergeSort(slow);
        head = merge(head, slow);
    }
    
    ListNode* sortList(ListNode* head) {
        mergeSort(head);
        return head;
    }
};

File: Day 090/Problem 148. Sort List.java
class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) {this.val = val;}
    ListNode(int val, ListNode next) {this.val = val; this.next = next;}
}

class Solution {

    private ListNode merge(ListNode left, ListNode right) {
        ListNode sentinel = new ListNode();
        ListNode ptr = sentinel;

        while (left != null && right != null) {
            if (left.val < right.val) {
                ptr.next = new ListNode(left.val);
                left = left.next;
            } else {
                ptr.next = new ListNode(right.val);
                right = right.next;
            }

            ptr = ptr.next;
        }

        if (left != null) ptr.next = left;
        if (right != null) ptr.next = right;

        return sentinel.next;
    }

    private ListNode mergeSort(ListNode head) {
        if (head == null || head.next == null) return head;

        // Find middle node
        ListNode prev = null, slow = head, fast = head;
        while (fast != null && fast.next != null) {
            prev = slow;
            slow = slow.next;
            fast = fast.next.next;
        }

        prev.next = null;
        ListNode left = mergeSort(head);
        ListNode right = mergeSort(slow);

        return merge(left, right);
    }

    public ListNode sortList(ListNode head) {
        return mergeSort(head);
    }
}